node/forest_libp2p/src/discovery.rs

```diff
@@ -215,55 +208,89 @@ impl DiscoveryBehaviour {
 }
 
 impl NetworkBehaviour for DiscoveryBehaviour {
-    type ConnectionHandler = ToggleIntoConnectionHandler<KademliaHandlerProto<QueryId>>;
-    type OutEvent = DiscoveryOut;
+    type ConnectionHandler = <KademliaBehaviour as NetworkBehaviour>::ConnectionHandler;
+    type OutEvent = DiscoveryEvent;
 
-    fn new_handler(&mut self) -> Self::ConnectionHandler {
-        self.kademlia.new_handler()
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        local_addr: &libp2p::Multiaddr,
+        remote_addr: &libp2p::Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.peer_addresses
+            .entry(peer)
+            .or_default()
+            .insert(remote_addr.clone());
+        self.kademlia.handle_established_inbound_connection(
+            connection_id,
+            peer,
+            local_addr,
+            remote_addr,
+        )
     }
 
-    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
-        let mut list = self
-            .user_defined
-            .iter()
-            .filter_map(|(p, a)| if p == peer_id { Some(a.clone()) } else { None })
-            .collect::<Vec<_>>();
-
-        {
-            let mut list_to_filter = Vec::new();
-            if let Some(k) = self.kademlia.as_mut() {
-                list_to_filter.extend(k.addresses_of_peer(peer_id))
-            }
-
-            list_to_filter.extend(self.mdns.addresses_of_peer(peer_id));
-
-            list.extend(list_to_filter);
-        }
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &libp2p::Multiaddr,
+        role_override: libp2p::core::Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.peer_addresses
+            .entry(peer)
+            .or_default()
+            .insert(addr.clone());
+        self.kademlia.handle_established_outbound_connection(
+            connection_id,
+            peer,
+            addr,
+            role_override,
+        )
+    }
 
-        trace!("Addresses of {:?}: {:?}", peer_id, list);
+    fn handle_pending_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        local_addr: &libp2p::Multiaddr,
+        remote_addr: &libp2p::Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        self.kademlia
+            .handle_pending_inbound_connection(connection_id, local_addr, remote_addr)
+    }
 
-        list
+    fn handle_pending_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        maybe_peer: Option<PeerId>,
+        addresses: &[libp2p::Multiaddr],
+        effective_role: libp2p::core::Endpoint,
+    ) -> Result<Vec<libp2p::Multiaddr>, ConnectionDenied> {
```

Comment:
To preserve the previous behaviour of `addresses_of_peer`, shouldn't you concatenate the `user_defined` addresses to the ones returned by Kademlia here?

Reply:
`user_defined` has been set on [initialization](https://github.com/ChainSafe/forest/pull/2598/files#diff-fa6b31f7c0d7e628fdb5238e78ee8a073b1e52ad1618fb1e08bfa0828651b250R130)